evm/src/cpu/kernel/optimizer.rs

```diff
@@ -0,0 +1,234 @@
+use ethereum_types::U256;
+use Item::{Push, StandardOp};
+use PushTarget::Literal;
+
+use crate::cpu::kernel::ast::Item::{GlobalLabelDeclaration, LocalLabelDeclaration};
+use crate::cpu::kernel::ast::PushTarget::Label;
+use crate::cpu::kernel::ast::{Item, PushTarget};
+use crate::cpu::kernel::utils::{replace_windows, u256_from_bool};
+
+pub(crate) fn optimize_asm(code: &mut Vec<Item>) {
+    // Run the optimizer until nothing changes.
+    loop {
+        let old_code = code.clone();
+        optimize_asm_once(code);
+        if code == &old_code {
+            break;
+        }
+    }
+}
+
+/// A single optimization pass.
+fn optimize_asm_once(code: &mut Vec<Item>) {
+    constant_propagation(code);
+    no_op_jumps(code);
+    remove_swapped_pushes(code);
+    remove_swaps_commutative(code);
+    remove_ignored_values(code);
+}
+
+/// Constant propagation.
+fn constant_propagation(code: &mut Vec<Item>) {
```

Comment:
One thing to keep in mind: constant propagation isn’t _guaranteed_ to be helpful. It depends on how we implement `PUSH` instructions.

In particular, we discussed implementing `PUSHn` as a system call (especially for large `n`). Consider the following code: `[PUSH1 0, NOT]`. It effectively pushes 2^256 − 1 onto the stack, so we could replace it with `[PUSH32 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff]`. But if `PUSH1` is one row while `PUSH32` is a system call, then the original version is actually faster and constant propagation actually does more harm than good.

Reply:
Good point. Maybe we should treat the optimization rules as suggestions that are passed through a cost estimator. We could suggest the opposite optimization too, `[PUSH x] -> [PUSH !x, NOT]`, which might help with some curve code.